api/lcp/collection.py

```diff
@@ -0,0 +1,311 @@
+        raise an exception.
+
+        :param delivery_mechanism: A LicensePoolDeliveryMechanism
+        :type delivery_mechanism: LicensePoolDeliveryMechanism
+        """
+        return delivery_mechanism
+
+    @property
+    def collection(self):
+        """Returns an associated Collection object
+
+        :return: Associated Collection object
+        :rtype: Collection
+        """
+        return Collection.by_id(self._db, id=self._collection_id)
+
+    def external_integration(self, db):
+        """Returns an external integration owned by this object
+
+        :param db: Database session
+        :type db: sqlalchemy.orm.session.Session
+
+        :return: External integration owned by this object
+        :rtype: core.model.configuration.ExternalIntegration
+        """
+        return self.collection.external_integration
+
+    def _create_lcp_server(self):
+        """Creates a new instance of LCPServer
+
+        :return: New instance of LCPServer
+        :rtype: LCPServer
+        """
+        configuration_storage = ConfigurationStorage(self)
+        configuration_factory = ConfigurationFactory()
+        hasher_factory = HasherFactory()
+        lcp_server = LCPServer(configuration_storage, configuration_factory, hasher_factory)
+
+        return lcp_server
+
+    @property
+    def _lcp_server(self):
+        """Returns an instance of LCPServer
+
+        :return: Instance of LCPServer
+        :rtype: LCPServer
+        """
+        if self._lcp_server_instance is None:
+            self._lcp_server_instance = self._create_lcp_server()
+
+        return self._lcp_server_instance
+
+    def checkout(self, patron, pin, licensepool, internal_format):
+        """Checks out a book on behalf of a patron
+
+        :param patron: A Patron object for the patron who wants to check out the book
+        :type patron: Patron
+
+        :param pin: The patron's alleged password
+        :type pin: string
+
+        :param licensepool: Contains lending info as well as link to parent Identifier
+        :type licensepool: LicensePool
+
+        :param internal_format: Represents the patron's desired book format.
+        :type internal_format: Any
+
+        :return: a LoanInfo object
+        :rtype: LoanInfo
+        """
+        days = self.collection.default_loan_period(patron.library)
+        today = datetime.datetime.utcnow()
+        expires = today + datetime.timedelta(days=days)
+        loan = get_one(self._db, Loan, patron=patron, license_pool=licensepool, on_multiple='interchangeable')
+
+        if loan:
+            license = self._lcp_server.get_license(self._db, loan.external_identifier)
+        else:
+            license = self._lcp_server.generate_license(
+                self._db,
+                licensepool.identifier.identifier,
+                patron.id,
+                today,
+                expires
+            )
+
+            patron.external_identifier = license['encryption']['user_key']['key_check']
```

Comment:
`patron.external_identifier` is used for the unique identifier assigned to the patron by the ILS, so we can track the patron across changes to their library card number. It's not appropriate to use it to store any kind of information from an LCP server. 

I'm not 100% sure what this value is for, but since it differs from one loan to another it would be appropriate to store it as `loan.external_identifier`. But you're using that for `license['id']`, which makes more sense.

Is it necessary to store this value at all?

Reply:
`license['encryption']['user_key']['key_check']` contains patron's hashed passphrase, I need to store it somewhere so I can later [add it to OPDS feed](https://github.com/NYPL-Simplified/server_core/pull/1194/files#diff-eb69c9549013acc45b217e06734d0ec9R1522)

Reply:
`loan.external_identifier` is already used to store license's ID which is needed to be able to fetch an existing license by its ID.
What would be the most appropriate place where I can store patron's hashed passphrase?

Reply:
Also, I'm thinking about what should be a passphrase. Currently I'm using `LCPServer.DEFAULT_PATRON_PASSPHRASE` because we rely on [automatic key retrieval feature](https://readium.org/lcp-specs/releases/lcp/latest.html#36-identifying-rights-and-restrictions-the-rights-object) and patrons should never enter their passphrases manually but on the other hand patrons will not be able to use other devices not supporting this feature. Also, it might be not very safe to use the same passphrase for all patrons and books. Does it make sense to generate it dynamically as a UUID for example?

Reply:
https://readium.org/lcp-specs/releases/lcp/latest.html#44-requirements-for-the-user-key--user-passphrase says: "the Provider should use the same User Key for all licenses issued to the same User." Since the User Key is derived from the User Passphrases I'm OK with having a single User Passphrase.

So, store the User Passphrase for a patron in a permanent `Credential` for that patron. The `Credential` should have data source DataSource.INTERNAL_PROCESSING and a new type, maybe `LCP Passphrase`. The value can be a UUID. 

Another way to do this is to use `Patron.authorization_identifier` as the User Passphrase. Since the user knows this value they'll be able to type it in in another reader.  But patrons do change their authorization_identifiers occasionally and I don't know if patrons who authenticate through SAML will have an authorization_identifier, so I think a UUID is safer.